packages/css/src/uma/UmaClient.ts

```diff
@@ -28,44 +24,154 @@ export type UmaClaims = JWTPayload & {
 
 export interface UmaConfig {
   jwks_uri: string;
-  jwks: any;
+  // jwks: any;
   issuer: string;
   permission_endpoint: string;
   introspection_endpoint: string;
   resource_registration_endpoint: string;
 }
 
-export type PermissionTicketRequest = {
-  ticketSubject: string,
-  owner: string,
-  ticketNeeds: Set<string>
+export type UmaVerificationOptions = Omit<JWTVerifyOptions, 'iss' | 'aud' | 'sub' | 'iat'>;
+
+const UMA_DISCOVERY = '/.well-known/uma2-configuration';
+
+const REQUIRED_METADATA = [
+  'issuer', 
+  'jwks_uri', 
+  'permission_endpoint', 
+  'introspection_endpoint', 
+  'resource_registration_endpoint'
+];
+
+const algMap = {
+  'ES256': { name: 'ECDSA', namedCurve: 'P-256', hash: 'SHA-256' },
+  'ES384': { name: 'ECDSA', namedCurve: 'P-384', hash: 'SHA-384' },
+  'ES512': { name: 'ECDSA', namedCurve: 'P-512', hash: 'SHA-512' },
+  'HS256': { name: 'HMAC', hash: 'SHA-256' },
+  'HS384': { name: 'HMAC', hash: 'SHA-384' },
+  'HS512': { name: 'HMAC', hash: 'SHA-512' },
+  'PS256': { name: 'RSASSA-PSS', hash: 'SHA-256' },
+  'PS384': { name: 'RSASSA-PSS', hash: 'SHA-384' },
+  'PS512': { name: 'RSASSA-PSS', hash: 'SHA-512' },
+  'RS256': { name: 'RSASSA-PKCS1-v1_5', hash: 'SHA-256' },
+  'RS384': { name: 'RSASSA-PKCS1-v1_5', hash: 'SHA-384' },
+  'RS512': { name: 'RSASSA-PKCS1-v1_5', hash: 'SHA-512' },
 }
 
 /**
  * Client interface for the UMA AS
  */
-export abstract class UmaClient {
-  protected abstract readonly logger: Logger;
-  protected abstract options: UmaVerificationOptions;
-  protected abstract retrievePat(issuer: string, owner: string): Promise<string>;
-    
+export class UmaClient {
+  protected readonly logger = getLoggerFor(this);
+
+  /**
+   * @param {JwkGenerator} keyGen - the generator providing the signing key
+   * @param {UmaVerificationOptions} options - options for JWT verification
+   */
+  constructor(
+    protected baseUrl: string,
+    protected keyGen: JwkGenerator, 
+    protected fetcher: Fetcher,
+    protected options: UmaVerificationOptions = {},
+  ) {}
+
+  public async signedFetch(url: string, request: RequestInit & Omit<SignRequest, 'url'>): Promise<Response> {
```

Comment:
Correct. Plans are not clear yet. UMA says PAT per owner, GNAP says HTTP Message Signatures between servers (similar to this implementation). Not sure yet about my preference. Any thoughts?

Reply:
Having it be server-side definitely seems easier, as otherwise users somehow have to be responsible for providing their own PAT. Unless they let the server generate it for them, but at that point you might as well have just one for the server. If you trust the server to store all your data, I don't really see the point in having a personal PAT.